File: Tree/Stack/98.validate-binary-search-tree.cpp
/*
 * @lc app=leetcode id=98 lang=cpp
 * 1. divide and conquer; left subtree return max; right return min
 * 2. inorder traversal: an increasing sequence
 * [98] Validate Binary Search Tree
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        //left-root-right Inorder;
        stack<TreeNode*> stack;
        vector<int> inorder;
        TreeNode* temp  = root;
        while(temp||!stack.empty()){
            while(temp){
                stack.push(temp);
                temp=temp->left;
            }
            temp=stack.top();
            stack.pop();
            inorder.push_back(temp->val);
            temp = temp->right;
        }
        return checkValid (inorder);
    }
    bool checkValid (vector<int> v){
        for(int i = 1; i<v.size();i++){
            if(v[i-1]>=v[i]) return false;
        }
        return true;
    }
};

File: Tree/Divide&Conquer/98.validate-binary-search-tree.java
/*
 * @lc app=leetcode id=98 lang=java
 * Divide and Conquer
 * [98] Validate Binary Search Tree
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        checkValid(root, null, null);
    }
    public boolean checkValid(TreeNode root, TreeNode min, TreeNode max){
        if(root==null) return true;
        if(min!=null && root.val<=min.val || max!=null && root.val>=max.val) return false;
        return checkValid(root.left, min, root) && checkValid(root.right, root, max);
    }
}

